src/background/services/tabEvents.ts

```diff
@@ -1,16 +1,55 @@
+import { type Browser, runtime } from 'webextension-polyfill'
 import { MonetizationService } from './monetization'
+import { StorageService } from './storage'
+import { IsTabMonetizedPayload } from '@/shared/messages'
 
-export class TabEvents {
-  constructor(private monetizationService: MonetizationService) {}
-
-  onRemovedTab = (tabId: number) => {
-    this.monetizationService.clearTabSessions(tabId)
-  }
+const iconActive34 = runtime.getURL('assets/icons/icon-active-34.png')
+const iconActive128 = runtime.getURL('assets/icons/icon-active-128.png')
+const iconInactive34 = runtime.getURL('assets/icons/icon-inactive-34.png')
+const iconInactive128 = runtime.getURL('assets/icons/icon-inactive-128.png')
+const iconWarning34 = runtime.getURL('assets/icons/icon-warning-34.png')
+const iconWarning128 = runtime.getURL('assets/icons/icon-warning-128.png')
 
+export class TabEvents {
+  constructor(
+    private monetizationService: MonetizationService,
+    private storage: StorageService,
+    private browser: Browser
+  ) {}
   // TODO: This is not ideal. Find a better way to clear the sessions for a specific tab.
   // When closing the tab, we receive the STOP_MONETIZATION message as well.
   // Maybe check if the tab is closed in the content script?
-  onUpdatedTab = (tabId: number) => {
+  onRemovedTab = (tabId: number) => {
     this.monetizationService.clearTabSessions(tabId)
   }
+
+  onUpdatedTab = async (payload?: IsTabMonetizedPayload) => {
+    let iconData = {
+      '34': iconInactive34,
+      '128': iconInactive128
+    }
+
+    if (payload) {
+      const { value } = payload
+
+      if (value)
+        iconData = {
+          '34': iconActive34,
+          '128': iconActive128
+        }
+    } else {
+      const { enabled } = await this.storage.get(['enabled'])
+
+      iconData = {
+        '34': enabled ? iconActive34 : iconWarning34,
+        '128': enabled ? iconActive128 : iconWarning128
+      }
```

Comment:
I believe you would have to check this first always, otherwise you can be on a page, toggle WM off (the inactive icon is shown), then navigate to a page, and the icon will be either the active or warn, ignoring the WM toggle.

Reply:
If I'm understanding correctly, I guess there is also a limitation here: if we have the WM toggle to off (warning icon),  and then toggle it on, we don't know if the current page is web monetized or not? Is this something that can be stored in `storage`?

Reply:
@mkurapov good catch. I changed the logic to signal the monetization to start & be reflected in the icon if the WM is enabled again on the page.